### DISCUSSION TITLE
cannot import name syncdetect

### DISCUSSION THREAD

**Discussion by ChewingPencilLines** [pyinstaller/pyinstaller#6550] [Help] [answered]
I successfully generated an executable (on Ubuntu) but when running it, I get this error.
Tried looking into this, but can't seem to get a handle on it.

These are the installed versions of my CUPY and CUDA in this minicoda virtual environment.

> CUPY: 9.6.0
> CUDA: 11.4 

The odd thing is that when I first got this error last night, I followed the suggested correction (using `pip install cupy-cuda114`) and reinstalled CUDA, and the error went away... However, it reappeared today. I am utterly confused. I tried reinstalling CUDA again using `conda instalal -c conda-forge cupy` but to no avail.

```
Traceback (most recent call last):
  File "cupy/__init__.py", line 18, in <module>
  File "PyInstaller/loader/pyimod03_importers.py", line 495, in exec_module
  File "cupy/_core/__init__.py", line 1, in <module>
  File "cupy/_core/core.pyx", line 1, in init cupy._core.core
  File "PyInstaller/loader/pyimod03_importers.py", line 495, in exec_module
  File "cupy/cuda/__init__.py", line 8, in <module>
  File "PyInstaller/loader/pyimod03_importers.py", line 495, in exec_module
  File "cupy/cuda/compiler.py", line 13, in <module>
  File "cupy/cuda/device.pyx", line 5, in init cupy.cuda.device
ImportError: cannot import name syncdetect

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "cli.py", line 1, in <module>
    from package_a.c_cluster import main
  File "PyInstaller/loader/pyimod03_importers.py", line 495, in exec_module
  File "package_a/c_cluster.py", line 50, in <module>
    import cudf as pd
  File "PyInstaller/loader/pyimod03_importers.py", line 495, in exec_module
  File "cudf/__init__.py", line 6, in <module>
  File "PyInstaller/loader/pyimod03_importers.py", line 495, in exec_module
  File "cupy/__init__.py", line 27, in <module>
ImportError:
================================================================
Failed to import CuPy.

If you installed CuPy via wheels (cupy-cudaXXX or cupy-rocm-X-X), make sure that the package matches with the version of CUDA or ROCm installed.

On Linux, you may need to set LD_LIBRARY_PATH environment variable depending on how you installed CUDA/ROCm.
On Windows, try setting CUDA_PATH environment variable.

Check the Installation Guide for details:
  https://docs.cupy.dev/en/latest/install.html

Original error:
  ImportError: cannot import name syncdetect
================================================================
```

Any help would be greatly appreciated! >_<

**Comment by rokm** [comment, marked as answer]
> File "cupy/cuda/device.pyx", line 5, in init cupy.cuda.device
> ImportError: cannot import name syncdetect

`cupy` consists of several compiled modules, which means that PyInstaller cannot analyze their imports. So those imports end up missing in the frozen application, and you need to manually ensure they are collected, for example using [--hidden-import](https://pyinstaller.readthedocs.io/en/stable/usage.html#cmdoption-hidden-import) (on per-module basis) or [--collect-submodules](https://pyinstaller.readthedocs.io/en/stable/usage.html#cmdoption-collect-submodules) (whole package, including all its subpackages and submodules) command-line switches (or their spec file equivalents).

For example, a basic import test:
```python
# program.py
import cupy
print(cupy)
```
needs to be frozen with: 
```
pyinstaller --clean --noconfirm program.py  \
    --collect-submodules cupy \
    --collect-submodules cupy_backends.cuda \
    --collect-submodules fastrlock
``` 

That's using `cupy-cuda115` PyPI wheel; no idea if there's anything special about conda's package...

**Reply by ChewingPencilLines** [reply]
Hi,

I managed to get past the above issue, thank you very much.
However, I am now getting the following error:

`Cannot find library libtreelite.so in the candidate path: List of candidates: /tmp/_MEIXPS1rq/treelite/lib/libtreelite.so`

I have attempted to add the following `hidden_imports` into the `cli.spec` file to no avail:

```
libtreelite.so
libtreelite
treelite
treelite.lib
treelite_runtime
```

I have also attempted to collecting submodules by adding the following, also to no avail:
`hiddenimports += collect_submodules('cuml')`

Any help would be greatly appreciated! >_<

**Reply by bwoodsend** [reply]
`libtreelite.so` is a C library - not a Python module so it's `--collect-binaries treelite` you want.

**Reply by ChewingPencilLines** [reply]
> `libtreelite.so` is a C library - not a Python module so it's `--collect-binaries treelite` you want.

I attempted adding `binaries = ["treelite"]` to Analysis in the `cli.spec` file and that didn't work.
Can you advise?

**Reply by bwoodsend** [reply]
If you're using the spec file then you need to use the corresponding spec file function [collect_dynamic_libs()](https://pyinstaller.readthedocs.io/en/v4.8/hooks.html#PyInstaller.utils.hooks.collect_dynamic_libs). `binaries = collect_dynamic_libs("treelite")`.

**Reply by ChewingPencilLines** [reply]
> If you're using the spec file then you need to use the corresponding spec file function [collect_dynamic_libs()](https://pyinstaller.readthedocs.io/en/v4.8/hooks.html#PyInstaller.utils.hooks.collect_dynamic_libs). `binaries = collect_dynamic_libs("treelite")`.

I managed to get past this and the subsequent errors regarding dynamic libraries, hidden imports and the likes.
However, when I tried to run the generated executable, I am getting the following error:

`File "treelite_runtime/__init__.py", line 8, in <module>
FileNotFoundError: [Errno 2] No such file or directory: '/mnt/store/ttt/tmp/_MEIxXzs86/treelite_runtime/VERSION'`

As this error is completely unknown to me, I tried the drastic solution of remaking an entirely new environment from scratch with a predefined `tmp` directory. However, this did not resolve the problem. The `tmp` folder does not contain the required file. It is unclear to me whether this is a problem related to `pyinstaller` or `treelite`. Can you help?

Many thanks for all the great replies and help rendered to me thus far. I really appreciate it.
If not for this help, I would not have been able to get this far.

**Reply by rokm** [reply]
> `File "treelite_runtime/__init__.py", line 8, in <module> FileNotFoundError: [Errno 2] No such file or directory: '/mnt/store/ttt/tmp/_MEIxXzs86/treelite_runtime/VERSION'`

This is a variant of missing files that you have already been solving, except that in this case, it is a data file (as opposed to python module or shared library) that's missing. So `--collect-data treelite` (or equivalent `datas = collect_data_files('treelite')` in the spec file) should do the trick.

**Reply by ChewingPencilLines** [reply]
> > `File "treelite_runtime/__init__.py", line 8, in <module> FileNotFoundError: [Errno 2] No such file or directory: '/mnt/store/ttt/tmp/_MEIxXzs86/treelite_runtime/VERSION'`
> 
> This is a variant of missing files that you have already been solving, except that in this case, it is a data file (as opposed to python module or shared library) that's missing. So `--collect-data treelite` (or equivalent `datas = collect_data_files('treelite')` in the spec file) should do the trick.

Wow, I can't seem to get a handle on how to classify these things.
Do you have tips on advice regarding this?
Instead of me coming here to get free help and bothering people with my questions, >_>

For example, I am now getting this:
`numba.cuda.cudadrv.error.CudaSupportError: Error at driver init: Call to cuInit results in CUDA_ERROR_OPERATING_SYSTEM (304)
[26063] Failed to execute script 'cli' due to unhandled exception!`

It doesn't look like the errors I encountered previously that required me to add to `hiddenimports`, `binaries`, `metadata`, and `datafiles`.
How do I tell?

Thanks and best!

_A git issue suggests that this has to do with symlink?_

**Reply by rokm** [reply]
> Wow, I can't seem to get a handle on how to classify these things.
> Do you have tips on advice regarding this?
> Instead of me coming here to get free help and bothering people with my questions, >_>

Unfortunately, this is the sort of thing that comes with practice. Once you see a lot of different errors, you start to notice patterns. But the general rule is, PyInstaller cannot detect indirectly imported python modules (which requires `--hiddenimport` or `--collect-submodules`), does not collect non-python data files at all (hence `--collect-data`), and may or may not collect shared libraries (depending on whether they are linked against extension modules or dlopen'd, and how search paths are set; hence `--collect-binaries`). There's also a problem that auto-detected shared libraries are collected into top-level application directory, which may sometimes cause issues.

> For example, I am now getting this:
> `numba.cuda.cudadrv.error.CudaSupportError: Error at driver init: Call to cuInit results in CUDA_ERROR_OPERATING_SYSTEM (304) [26063] Failed to execute script 'cli' due to unhandled exception!`

This could be anything, really. A missing shared library, an incorrectly placed shared library, etc.

> _A git issue suggests that this has to do with symlink?_

Could be. PyInstaller does currently not support symlinks, so it will create distinct copies of the file instead of symbolic links. There's currently no way to work around that, except by manually removing the extra copies and replacing them with symlinks (either manually or as part of post-processing steps, if you use an automated/scripted build process).

Can you check if there are multiple copies of libcuda.so.* collected in the application directory, and try replacing the extra copies with symlinks?

**Reply by ChewingPencilLines** [reply]
> application directory

I am not able to find an application directory after invoking the `pyinstaller` command with the `--onefile` flag.

In the created `dist` folder, I have a `cli` file, and a `stpFhtfR` file.
In the created `build` folder, I have a `cli` folder containing multiple `.toc` files and a `base_library.zip`, as well as a `PYZ-00.pyz` file and a `xref-cli.html`.

As such, I am unsure of which application directory you are referring to.

Can you please advise?

**Reply by rokm** [reply]
Yeah, manually restoring symlinks won't work with `onefile`. You'll have to use `onedir`.

**Reply by rokm** [reply]
And if the problem are indeed duplicated libs and lack of symlinks, you can later work-around it in `onefile` bundle by removing and relinking the offending files at the start of your program (before you import any of the cuda-based packages).

**Reply by ChewingPencilLines** [reply]
> Yeah, manually restoring symlinks won't work with `onefile`. You'll have to use `onedir`.

As advised, I tried running pyinstaller with `onedir`.

Using `find . -name "libcuda.so.*"` in the directory `'/mnt/store/ttt/'` (basically where `cli.spec` is located), only one result was located:
`./dist/cli/libcuda.so.1`.

Does this mean that it is not an issue with duplicate copies of `libcuda` and therefore not an issue with symlinks?
If so, I must confess that I am completely lost and have no idea how to proceed, T_T.

**Reply by rokm** [reply]
> Using `find . -name "libcuda.so.*"` in the directory `'/mnt/store/ttt/'` (basically where `cli.spec` is located), only one result was located:
> `./dist/cli/libcuda.so.1`.

What about `libcu*.so*`?

Can you check in `build/cli/Analysis-00.toc` where `libcuda.so.1` was collected from?

That said, does this application use multiple CUDA-based libraries, e.g., `cupy` *and* `numba.cuda`? If that's the case, how was `cupy` installed, via `conda` or via `pip`? 

I suggest you try freezing a simple `numba.cuda` based example, and see if you can get that working. 

<details>
<summary>Example from https://nyu-cds.github.io/python-numba/05-cuda</summary>
<p>

```python
from __future__ import division
from numba import cuda
import numpy
import math

# CUDA kernel
@cuda.jit
def my_kernel(io_array):
    pos = cuda.grid(1)
    if pos < io_array.size:
        io_array[pos] *= 2 # do the computation

# Host code   
data = numpy.ones(256)
threadsperblock = 256
blockspergrid = math.ceil(data.shape[0] / threadsperblock)
my_kernel[blockspergrid, threadsperblock](data)
print(data)
``` 

</p>
</details>

I tried here, and it looks like no CUDA libraries are actually collected from the conda environment, and to get it working, I had to ensure that `libnvvm.so` and `libdevice.10.bc` are collected to `lib` sub-directory in application's directory (looks like `numba.cuda` is very particular about its search paths). And even then, its JIT compiler actually made use of CUDA libraries that were installed on my base system, outside of the application bundle (so maybe that's why other libraries did not need to be collected)... 

All in all, in my experience deploying CUDA-based applications is complex enough on its own (assuming CUDA is not available in the target environment and needs to be bundled), which is made even trickier by trying to deploy it as part of python-based frameworks via PyInstaller. So I think this falls well out of scope of the help we can provide here, especially without seeing the application code and being able to see what's going on.
